internal/lock.go

```diff
@@ -50,160 +49,199 @@ func NewLock(path string, newOk bool) (*Lock, error) {
+}
+
 func (l *Lock) DeleteResource(path string) {
-	newStatements := []Resource{}
+	log.Debug().Str("path", path).Msg("Deleting resource")
+
+	newResources := []Resource{}
 	for _, r := range l.conf.Resource {
 		if !r.Contains(path) {
-			newStatements = append(newStatements, r)
+			newResources = append(newResources, r)
 		}
 	}
-	l.conf.Resource = newStatements
+
+	removed := len(l.conf.Resource) - len(newResources)
+	l.conf.Resource = newResources
+
+	log.Debug().Int("removed", removed).Msg("Resources deleted")
 }
 
 const NoFileMode = os.FileMode(0)
 
-// strToFileMode converts a string to a os.FileMode.
 func strToFileMode(perm string) (os.FileMode, error) {
 	if perm == "" {
 		return NoFileMode, nil
 	}
 	parsed, err := strconv.ParseUint(perm, 8, 32)
 	if err != nil {
-		return NoFileMode, err
+		return NoFileMode, fmt.Errorf("invalid permission format: %w", err)
 	}
 	return os.FileMode(parsed), nil
 }
 
-// Download gets all the resources in this lock file and moves them to
-// the destination directory.
 func (l *Lock) Download(dir string, tags []string, notags []string, perm string) error {
 	if stat, err := os.Stat(dir); err != nil || !stat.IsDir() {
 		return fmt.Errorf("'%s' is not a directory", dir)
 	}
+
 	mode, err := strToFileMode(perm)
 	if err != nil {
-		return fmt.Errorf("'%s' is not a valid permission definition", perm)
+		return err
+	}
+
+	resources := l.filterResources(tags, notags)
+	if len(resources) == 0 {
+		return fmt.Errorf("no resources to download")
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	// Filter in the resources that have all the required tags.
-	tagFilteredResources := []Resource{}
-	if len(tags) > 0 {
-		for _, r := range l.conf.Resource {
-			hasAllTags := true
-			for _, tag := range tags {
-				hasTag := false
-				for _, rtag := range r.Tags {
-					if tag == rtag {
-						hasTag = true
-						break
-					}
-				}
-				if !hasTag {
-					hasAllTags = false
-					break
-				}
-			}
-			if hasAllTags {
-				tagFilteredResources = append(tagFilteredResources, r)
-			}
+
+	return l.downloadResources(resources, dir, mode, ctx)
+}
+
+func (l *Lock) filterResources(tags, notags []string) []Resource {
+	if len(tags) == 0 && len(notags) == 0 {
+		return l.conf.Resource
+	}
+
+	var filtered []Resource
+	for _, r := range l.conf.Resource {
+		if hasAllTags(r.Tags, tags) && !hasAnyTag(r.Tags, notags) {
+			filtered = append(filtered, r)
 		}
-	} else {
-		tagFilteredResources = l.conf.Resource
-	}
-	// Filter out the resources that have any 'notag' tag.
-	filteredResources := []Resource{}
-	if len(notags) > 0 {
-		for _, r := range tagFilteredResources {
-			hasTag := false
-			for _, notag := range notags {
-				for _, rtag := range r.Tags {
-					if notag == rtag {
-						hasTag = true
-					}
-				}
-			}
-			if !hasTag {
-				filteredResources = append(filteredResources, r)
+	}
+	return filtered
+}
+
+func hasAllTags(resourceTags, requiredTags []string) bool {
```

Comment:
Why are you refactoring this? I'm not totally opposed to it, but I think this PR will be complex enough to *not* do anything other than the task at hand.